pkg/apis/seedmanagement/v1alpha1/defaults.go

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	"encoding/json"
+	"fmt"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	"github.com/gardener/gardener/pkg/apis/seedmanagement/helper"
+	configv1alpha1 "github.com/gardener/gardener/pkg/gardenlet/apis/config/v1alpha1"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+func addDefaultingFuncs(scheme *runtime.Scheme) error {
+	return RegisterDefaults(scheme)
+}
+
+// SetDefaults_ManagedSeed sets default values for ManagedSeed objects.
+func SetDefaults_ManagedSeed(obj *ManagedSeed) {
+	switch {
+	case obj.Spec.SeedTemplate != nil:
+		// Set seed defaults
+		seed := gardencorev1beta1.Seed{Spec: obj.Spec.SeedTemplate.Spec}
+		gardencorev1beta1.SetDefaults_Seed(&seed)
+		obj.Spec.SeedTemplate.Spec = seed.Spec
```

Comment:
The deployment defaults **do** default the included pod spec, and not only deployment, all other Kubernetes types, see https://github.com/kubernetes/kubernetes/blob/6ba09e9a525225430f43233f96fb0c553f6ea1d2/pkg/apis/core/validation/validation.go#L4571. I explicitly checked this, and from my perspective it makes a lot of sense to do this here as well.

Reply:
Actually, after discussing the "merge with parent" with @rfranzke , I am not sure anymore in what I wrote above. To avoid the "merge with parent" issue, we decided to not default the `GardenletConfiguration` resource (since it breaks merging with parent), but to keep validation (since we do really need it). To enable this, I will change the validation code to "decouple" it from the defaults, e.g. making sure an empty `GardenletConfiguration` resource can pass validation.

Regarding the `Seed` resource above, we don't have such an issue, validation of an empty seed passes already today. So it's really a matter of what we decide to be better. For consistency with `GardenletConfiguration` I tend to agree that it's better to not default the `Seed`. Do you have other considerations besides this?

Reply:
> To enable this, I will change the validation code to "decouple" it from the defaults, e.g. making sure an empty GardenletConfiguration resource can pass validation.

This sounds like a good plan to me.

> Do you have other considerations besides this?

Hmm, currently not. If some concerns arise, I will add them to the new PR and we can discuss them there.

Reply:
OK, I will remove defaults on all embedded resources, including `Seed`.